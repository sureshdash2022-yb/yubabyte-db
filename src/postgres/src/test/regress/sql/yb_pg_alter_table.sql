--
-- ALTER_TABLE
--

-- Clean up in case a prior regression run failed
SET client_min_messages TO 'warning';
DROP ROLE IF EXISTS regress_alter_table_user1;
RESET client_min_messages;

CREATE USER regress_alter_table_user1;

--
-- rename - check on both non-temp and temp tables
--
CREATE TABLE attmp (regtable int);
CREATE TEMP TABLE attmp (attmptable int);

ALTER TABLE attmp RENAME TO attmp_new;

SELECT * FROM attmp;
SELECT * FROM attmp_new;

ALTER TABLE attmp RENAME TO attmp_new2;

SELECT * FROM attmp;		-- should fail
SELECT * FROM attmp_new;
SELECT * FROM attmp_new2;

DROP TABLE attmp_new;
DROP TABLE attmp_new2;

-- check rename of partitioned tables and indexes also
CREATE TABLE part_attmp (a int primary key) partition by range (a);
CREATE TABLE part_attmp1 PARTITION OF part_attmp FOR VALUES FROM (0) TO (100);
ALTER INDEX part_attmp_pkey RENAME TO part_attmp_index;
ALTER INDEX part_attmp1_pkey RENAME TO part_attmp1_index;
ALTER TABLE part_attmp RENAME TO part_at2tmp;
ALTER TABLE part_attmp1 RENAME TO part_at2tmp1;
SET ROLE regress_alter_table_user1;
ALTER INDEX part_attmp_index RENAME TO fail;
ALTER INDEX part_attmp1_index RENAME TO fail;
ALTER TABLE part_at2tmp RENAME TO fail;
ALTER TABLE part_at2tmp1 RENAME TO fail;
RESET ROLE;
DROP TABLE part_at2tmp;

-- ALTER TABLE ... RENAME on non-table relations
-- renaming indexes (FIXME: this should probably test the index's functionality)
ALTER INDEX IF EXISTS __onek_unique1 RENAME TO attmp_onek_unique1;
ALTER INDEX IF EXISTS __attmp_onek_unique1 RENAME TO onek_unique1;

ALTER INDEX onek_unique1 RENAME TO attmp_onek_unique1;
ALTER INDEX attmp_onek_unique1 RENAME TO onek_unique1;

SET ROLE regress_alter_table_user1;
ALTER INDEX onek_unique1 RENAME TO fail;  -- permission denied
RESET ROLE;

-- renaming index should rename constraint as well
ALTER TABLE onek ADD CONSTRAINT onek_unique1_constraint UNIQUE (unique1);
ALTER INDEX onek_unique1_constraint RENAME TO onek_unique1_constraint_foo;
ALTER TABLE onek DROP CONSTRAINT onek_unique1_constraint_foo;

--
-- lock levels
--
drop type lockmodes;
create type lockmodes as enum (
 'SIReadLock'
,'AccessShareLock'
,'RowShareLock'
,'RowExclusiveLock'
,'ShareUpdateExclusiveLock'
,'ShareLock'
,'ShareRowExclusiveLock'
,'ExclusiveLock'
,'AccessExclusiveLock'
);

drop view my_locks;
create or replace view my_locks as
select case when c.relname like 'pg_toast%' then 'pg_toast' else c.relname end, max(mode::lockmodes) as max_lockmode
from pg_locks l join pg_class c on l.relation = c.oid
where virtualtransaction = (
        select virtualtransaction
        from pg_locks
        where transactionid = txid_current()::integer)
and locktype = 'relation'
and relnamespace != (select oid from pg_namespace where nspname = 'pg_catalog')
and c.relname != 'my_locks'
group by c.relname;

create table alterlock (f1 int primary key, f2 text);
insert into alterlock values (1, 'foo');
create table alterlock2 (f3 int primary key, f1 int);
insert into alterlock2 values (1, 1);

begin; alter table alterlock alter column f2 set statistics 150;
select * from my_locks order by 1;
rollback;

begin; alter table alterlock cluster on alterlock_pkey;
select * from my_locks order by 1;
commit;

begin; alter table alterlock set without cluster;
select * from my_locks order by 1;
commit;

begin; alter table alterlock set (fillfactor = 100);
select * from my_locks order by 1;
commit;

begin; alter table alterlock reset (fillfactor);
select * from my_locks order by 1;
commit;

begin; alter table alterlock set (toast.autovacuum_enabled = off);
select * from my_locks order by 1;
commit;

begin; alter table alterlock set (autovacuum_enabled = off);
select * from my_locks order by 1;
commit;

begin; alter table alterlock alter column f2 set (n_distinct = 1);
select * from my_locks order by 1;
rollback;

-- test that mixing options with different lock levels works as expected
begin; alter table alterlock set (autovacuum_enabled = off, fillfactor = 80);
select * from my_locks order by 1;
commit;

begin; alter table alterlock alter column f2 set storage extended;
select * from my_locks order by 1;
rollback;

-- TODO(jason): uncomment when doing issue #9106
-- begin; alter table alterlock alter column f2 set default 'x';
-- select * from my_locks order by 1;
-- rollback;

--
-- alter object set schema
--

create schema alter1;
create schema alter2;

create text search parser alter1.prs(start = prsd_start, gettoken = prsd_nexttoken, end = prsd_end, lextypes = prsd_lextype);
create text search configuration alter1.cfg(parser = alter1.prs);

alter text search parser alter1.prs set schema alter2;
alter text search configuration alter1.cfg set schema alter2;

-- clean up
drop schema alter2 cascade;

--
-- typed tables: OF / NOT OF
--

CREATE TYPE tt_t0 AS (z inet, x int, y numeric(8,2));
ALTER TYPE tt_t0 DROP ATTRIBUTE z;
CREATE TABLE tt0 (x int NOT NULL, y numeric(8,2));	-- OK
CREATE TABLE tt1 (x int, y bigint);					-- wrong base type
CREATE TABLE tt2 (x int, y numeric(9,2));			-- wrong typmod
CREATE TABLE tt3 (y numeric(8,2), x int);			-- wrong column order
CREATE TABLE tt4 (x int);							-- too few columns
CREATE TABLE tt5 (x int, y numeric(8,2), z int);	-- too few columns
CREATE TABLE tt6 () INHERITS (tt0);					-- can't have a parent
CREATE TABLE tt7 (x int, q text, y numeric(8,2)) WITH OIDS;
ALTER TABLE tt7 DROP q;								-- OK

ALTER TABLE tt0 OF tt_t0;
ALTER TABLE tt1 OF tt_t0;
ALTER TABLE tt2 OF tt_t0;
ALTER TABLE tt3 OF tt_t0;
ALTER TABLE tt4 OF tt_t0;
ALTER TABLE tt5 OF tt_t0;
ALTER TABLE tt6 OF tt_t0;
ALTER TABLE tt7 OF tt_t0;

CREATE TYPE tt_t1 AS (x int, y numeric(8,2));
ALTER TABLE tt7 OF tt_t1;			-- reassign an already-typed table
ALTER TABLE tt7 NOT OF;
\d tt7

