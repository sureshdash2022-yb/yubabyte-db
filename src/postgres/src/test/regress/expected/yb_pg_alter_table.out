--
-- ALTER_TABLE
--
-- Clean up in case a prior regression run failed
SET client_min_messages TO 'warning';
DROP ROLE IF EXISTS regress_alter_table_user1;
RESET client_min_messages;
CREATE USER regress_alter_table_user1;
--
-- rename - check on both non-temp and temp tables
--
CREATE TABLE attmp (regtable int);
CREATE TEMP TABLE attmp (attmptable int);
ALTER TABLE attmp RENAME TO attmp_new;
SELECT * FROM attmp;
 regtable
----------
(0 rows)

SELECT * FROM attmp_new;
 attmptable
------------
(0 rows)

ALTER TABLE attmp RENAME TO attmp_new2;
SELECT * FROM attmp;		-- should fail
ERROR:  relation "attmp" does not exist
LINE 1: SELECT * FROM attmp;
                      ^
SELECT * FROM attmp_new;
 attmptable
------------
(0 rows)

SELECT * FROM attmp_new2;
 regtable
----------
(0 rows)

DROP TABLE attmp_new;
DROP TABLE attmp_new2;
-- check rename of partitioned tables and indexes also
CREATE TABLE part_attmp (a int primary key) partition by range (a);
CREATE TABLE part_attmp1 PARTITION OF part_attmp FOR VALUES FROM (0) TO (100);
ALTER INDEX part_attmp_pkey RENAME TO part_attmp_index;
ALTER INDEX part_attmp1_pkey RENAME TO part_attmp1_index;
ALTER TABLE part_attmp RENAME TO part_at2tmp;
ALTER TABLE part_attmp1 RENAME TO part_at2tmp1;
SET ROLE regress_alter_table_user1;
ALTER INDEX part_attmp_index RENAME TO fail;
ERROR:  must be owner of index part_attmp_index
ALTER INDEX part_attmp1_index RENAME TO fail;
ERROR:  must be owner of index part_attmp1_index
ALTER TABLE part_at2tmp RENAME TO fail;
ERROR:  must be owner of table part_at2tmp
ALTER TABLE part_at2tmp1 RENAME TO fail;
ERROR:  must be owner of table part_at2tmp1
RESET ROLE;
DROP TABLE part_at2tmp;
-- ALTER TABLE ... RENAME on non-table relations
-- renaming indexes (FIXME: this should probably test the index's functionality)
ALTER INDEX IF EXISTS __onek_unique1 RENAME TO attmp_onek_unique1;
NOTICE:  relation "__onek_unique1" does not exist, skipping
ALTER INDEX IF EXISTS __attmp_onek_unique1 RENAME TO onek_unique1;
NOTICE:  relation "__attmp_onek_unique1" does not exist, skipping
ALTER INDEX onek_unique1 RENAME TO attmp_onek_unique1;
ALTER INDEX attmp_onek_unique1 RENAME TO onek_unique1;
SET ROLE regress_alter_table_user1;
ALTER INDEX onek_unique1 RENAME TO fail;  -- permission denied
ERROR:  must be owner of index onek_unique1
RESET ROLE;
-- renaming index should rename constraint as well
ALTER TABLE onek ADD CONSTRAINT onek_unique1_constraint UNIQUE (unique1);
ALTER INDEX onek_unique1_constraint RENAME TO onek_unique1_constraint_foo;
ALTER TABLE onek DROP CONSTRAINT onek_unique1_constraint_foo;
--
-- lock levels
--
drop type lockmodes;
ERROR:  type "lockmodes" does not exist
create type lockmodes as enum (
 'SIReadLock'
,'AccessShareLock'
,'RowShareLock'
,'RowExclusiveLock'
,'ShareUpdateExclusiveLock'
,'ShareLock'
,'ShareRowExclusiveLock'
,'ExclusiveLock'
,'AccessExclusiveLock'
);
drop view my_locks;
ERROR:  view "my_locks" does not exist
create or replace view my_locks as
select case when c.relname like 'pg_toast%' then 'pg_toast' else c.relname end, max(mode::lockmodes) as max_lockmode
from pg_locks l join pg_class c on l.relation = c.oid
where virtualtransaction = (
        select virtualtransaction
        from pg_locks
        where transactionid = txid_current()::integer)
and locktype = 'relation'
and relnamespace != (select oid from pg_namespace where nspname = 'pg_catalog')
and c.relname != 'my_locks'
group by c.relname;
create table alterlock (f1 int primary key, f2 text);
insert into alterlock values (1, 'foo');
create table alterlock2 (f3 int primary key, f1 int);
insert into alterlock2 values (1, 1);
begin; alter table alterlock alter column f2 set statistics 150;
ERROR:  ALTER TABLE ALTER column not supported yet
LINE 1: alter table alterlock alter column f2 set statistics 150;
                              ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
select * from my_locks order by 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
rollback;
begin; alter table alterlock cluster on alterlock_pkey;
ERROR:  ALTER TABLE CLUSTER not supported yet
LINE 1: alter table alterlock cluster on alterlock_pkey;
                              ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
select * from my_locks order by 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
commit;
begin; alter table alterlock set without cluster;
ERROR:  ALTER TABLE SET WITHOUT CLUSTER not supported yet
LINE 1: alter table alterlock set without cluster;
                              ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
select * from my_locks order by 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
commit;
begin; alter table alterlock set (fillfactor = 100);
ERROR:  ALTER TABLE SET not supported yet
LINE 1: alter table alterlock set (fillfactor = 100);
                              ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
select * from my_locks order by 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
commit;
begin; alter table alterlock reset (fillfactor);
ERROR:  ALTER TABLE RESET not supported yet
LINE 1: alter table alterlock reset (fillfactor);
                              ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
select * from my_locks order by 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
commit;
begin; alter table alterlock set (toast.autovacuum_enabled = off);
ERROR:  ALTER TABLE SET not supported yet
LINE 1: alter table alterlock set (toast.autovacuum_enabled = off);
                              ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
select * from my_locks order by 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
commit;
begin; alter table alterlock set (autovacuum_enabled = off);
ERROR:  ALTER TABLE SET not supported yet
LINE 1: alter table alterlock set (autovacuum_enabled = off);
                              ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
select * from my_locks order by 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
commit;
begin; alter table alterlock alter column f2 set (n_distinct = 1);
ERROR:  ALTER TABLE ALTER column not supported yet
LINE 1: alter table alterlock alter column f2 set (n_distinct = 1);
                              ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
select * from my_locks order by 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
rollback;
-- test that mixing options with different lock levels works as expected
begin; alter table alterlock set (autovacuum_enabled = off, fillfactor = 80);
ERROR:  ALTER TABLE SET not supported yet
LINE 1: alter table alterlock set (autovacuum_enabled = off, fillfac...
                              ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
select * from my_locks order by 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
commit;
begin; alter table alterlock alter column f2 set storage extended;
ERROR:  ALTER TABLE ALTER column not supported yet
LINE 1: alter table alterlock alter column f2 set storage extended;
                              ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
select * from my_locks order by 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
rollback;
-- TODO(jason): uncomment when doing issue #9106
-- begin; alter table alterlock alter column f2 set default 'x';
-- select * from my_locks order by 1;
-- rollback;
--
-- alter object set schema
--
create schema alter1;
create schema alter2;
create text search parser alter1.prs(start = prsd_start, gettoken = prsd_nexttoken, end = prsd_end, lextypes = prsd_lextype);
create text search configuration alter1.cfg(parser = alter1.prs);
alter text search parser alter1.prs set schema alter2;
alter text search configuration alter1.cfg set schema alter2;
-- clean up
drop schema alter2 cascade;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to text search configuration alter2.cfg
drop cascades to text search parser alter2.prs
--
-- typed tables: OF / NOT OF
--
CREATE TYPE tt_t0 AS (z inet, x int, y numeric(8,2));
ALTER TYPE tt_t0 DROP ATTRIBUTE z;
ERROR:  ALTER TYPE DROP ATTRIBUTE not supported yet
LINE 1: ALTER TYPE tt_t0 DROP ATTRIBUTE z;
                         ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1893. Click '+' on the description to raise its priority
CREATE TABLE tt0 (x int NOT NULL, y numeric(8,2));	-- OK
CREATE TABLE tt1 (x int, y bigint);					-- wrong base type
CREATE TABLE tt2 (x int, y numeric(9,2));			-- wrong typmod
CREATE TABLE tt3 (y numeric(8,2), x int);			-- wrong column order
CREATE TABLE tt4 (x int);							-- too few columns
CREATE TABLE tt5 (x int, y numeric(8,2), z int);	-- too few columns
CREATE TABLE tt6 () INHERITS (tt0);					-- can't have a parent
ERROR:  INHERITS not supported yet
LINE 1: CREATE TABLE tt6 () INHERITS (tt0);
                            ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1129. Click '+' on the description to raise its priority
CREATE TABLE tt7 (x int, q text, y numeric(8,2)) WITH OIDS;
ERROR:  OIDs are not supported for user tables.
ALTER TABLE tt7 DROP q;								-- OK
ERROR:  relation "tt7" does not exist
ALTER TABLE tt0 OF tt_t0;
ERROR:  ALTER TABLE OF not supported yet
LINE 1: ALTER TABLE tt0 OF tt_t0;
                        ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
ALTER TABLE tt1 OF tt_t0;
ERROR:  ALTER TABLE OF not supported yet
LINE 1: ALTER TABLE tt1 OF tt_t0;
                        ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
ALTER TABLE tt2 OF tt_t0;
ERROR:  ALTER TABLE OF not supported yet
LINE 1: ALTER TABLE tt2 OF tt_t0;
                        ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
ALTER TABLE tt3 OF tt_t0;
ERROR:  ALTER TABLE OF not supported yet
LINE 1: ALTER TABLE tt3 OF tt_t0;
                        ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
ALTER TABLE tt4 OF tt_t0;
ERROR:  ALTER TABLE OF not supported yet
LINE 1: ALTER TABLE tt4 OF tt_t0;
                        ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
ALTER TABLE tt5 OF tt_t0;
ERROR:  ALTER TABLE OF not supported yet
LINE 1: ALTER TABLE tt5 OF tt_t0;
                        ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
ALTER TABLE tt6 OF tt_t0;
ERROR:  ALTER TABLE OF not supported yet
LINE 1: ALTER TABLE tt6 OF tt_t0;
                        ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
ALTER TABLE tt7 OF tt_t0;
ERROR:  ALTER TABLE OF not supported yet
LINE 1: ALTER TABLE tt7 OF tt_t0;
                        ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
CREATE TYPE tt_t1 AS (x int, y numeric(8,2));
ALTER TABLE tt7 OF tt_t1;			-- reassign an already-typed table
ERROR:  ALTER TABLE OF not supported yet
LINE 1: ALTER TABLE tt7 OF tt_t1;
                        ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
ALTER TABLE tt7 NOT OF;
ERROR:  ALTER TABLE NOT OF not supported yet
LINE 1: ALTER TABLE tt7 NOT OF;
                        ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
\d tt7
